File: Inteligencia_Artificial_Programa_2/Puzzle/src/Platform.java
import java.util.*;

/**
 * Classe Platform: Esta classe contém construtores que criam objetos ContainerConfiguration
 * e o seu principal objetivo é criar sucessores de uma certa configurção para poder auxiliar
 * o método solve da classe BestFirst (Best First Search)
 * @author Márcio Felício
 * @version 1.0
 * @inv Cada Contêiner tem de ter um custo associado caso seja configuração inicial
 */
public class Platform implements Ilayout, Cloneable {
    /**
     * Estrutura estática da classe ContainerOrganizer
     */
    public static final Comparator<Stack<Container>> compareChars = (s1,s2) -> String.valueOf(
            s1.firstElement().getId()).compareTo(String.valueOf(s2.firstElement().getId()));

    /**
     * Estrutura da classe ContainerOrganizer
     */
    private List<Stack<Container>> stacks;
    private double cost;
    private Map<Character, int[]> cachedGoalPositions;
     /**
     * Construtor por omissão
     */
    public Platform() {
        stacks = null;
        this.cost = 0;
    }

    /**
     * Construtor de inicialização
     */
    public Platform(String config, boolean isInitialState) {
        stacks = new ArrayList<>();
        parseInput(config, isInitialState);
        this.cost = 0;
    }

    /**
     * Construtor de cópia para casos de encapsulamento de dados
     */
    public Platform(Platform other) {
        this.stacks = new ArrayList<>(other.stacks.size());

        for (Stack<Container> stack : other.stacks) {
            Stack<Container> newStack = new Stack<>();
            for (Container container : stack) {
                newStack.push(new Container(container));
            }
            this.stacks.add(newStack);
        }

        this.cost = other.getK();
    }

    private void parseInput(String config, boolean isInitialState) {
        String[] stacksConfig = config.split(" ");
        for (String stackStr : stacksConfig) {
            Stack<Container> stack = new Stack<>();
            for (int i = 0; i < stackStr.length(); i++) {
                char containerId = stackStr.charAt(i);

                int movecost = (isInitialState && i + 1 < stackStr.length() && Character.isDigit(stackStr.charAt(i + 1)))
                        ? Character.getNumericValue(stackStr.charAt(i + 1))
                        : 0;

                stack.push(new Container(containerId, movecost));

                if (movecost > 0) {
                    i++;
                }
            }
            stacks.add(stack);
        }
    }

    /**
     * getter para as stacks e setter para o custo de movimento de um contentor
     * @return stacks
     */
    public List<Stack<Container>> getStacks()
    {
        return stacks;
    }


    public void setCost(double cost)
    {
        this.cost = cost;
    }

    /**
     * @return uma chave única para a instância criada
     */
    @Override
    public int hashCode() {
        int hash = 7;

        for (Stack<Container> stack : stacks) {
            hash = 31 * hash + (stack.isEmpty() ? 0 : stack.hashCode());
        }

        return hash;
    }


    /**
     * @return um clone de uma determinada instância com o auxilio do construtor de cópia
     */
    @Override
    public Platform clone() {
        return new Platform(this);
    }

    /**
     * @return representação textual da estrutura da instância
     */
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();

        getSortedStacks();

        for (Stack<Container> stack : stacks) {
            sb.append("[");
            for (int i = 0; i < stack.size(); i++) {
                sb.append(stack.get(i).toString());
                if (i < stack.size() - 1) {
                    sb.append(", ");
                }
            }
            sb.append("]\n");
        }
        return sb.toString();
    }

    /**
     * @return as configurações feitas através de uma configuração pai
     */

    @Override
    public List<Ilayout> children() {
        List<Ilayout> children = new ArrayList<>();

        if (!this.stacks.isEmpty()) {
            for (int i = 0; i < this.stacks.size(); i++) {
                Stack<Container> sourceStack = this.stacks.get(i);

                for (int j = 0; j < this.stacks.size(); j++) {
                    if (i == j) continue;

                    Platform child = this.clone();

                    child.moveToStack(i,j);

                    if (child.stacks.get(i).isEmpty()) {
                        child.stacks.remove(i);
                    }

                    children.add(child);
                }

                if (sourceStack.size() >= 2) {
                    Platform child = this.clone();

                    child.moveToGround(i);

                    children.add(child);
                }
            }
        }
        return children;
    }

    /**
     * verifica se a configuração l é a configuração objetivo
     * @param l representa a configuração objetivo que queremos testar
     * @return true se forem iguais e false caso contrário
     */
    @Override
    public boolean isGoal(Ilayout l) {
        return this.equals(l);
    }

    /**
     * retorna o custo de movimento de um certo contentor
     */
    @Override
    public double getK() {
        return this.cost;
    }

    /**
     * Verifica se as configurações são iguais
     * @param o representa uma configuração que será testada com o this (instância receptora)
     * @return true caso sejam iguais e false caso contrário
     */
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Platform other = (Platform) o;

        this.getSortedStacks();
        other.getSortedStacks();

        if (this.stacks.size() != other.stacks.size()) return false;

        for (int i = 0; i < this.stacks.size(); i++) {
            Stack<Container> thisStack = this.stacks.get(i);
            Stack<Container> otherStack = other.stacks.get(i);

            if(thisStack.size() != otherStack.size()) return false;

            for (int j = 0; j < thisStack.size(); j++) {
                Container thisContainer = thisStack.get(j);
                Container goalContainer = otherStack.get(j);

                if (!thisContainer.equals(goalContainer)) {
                    return false;
                }
            }
        }
        return true;
    }




    /**
     * Ordena as stacks segundo o id dos contêiners, ignorando as stacks vazias.
     * @return lista de configurações ordenadas
     */
    private void getSortedStacks() {
        getStacks().sort(Comparator.comparing(stack -> stack.firstElement().getId()));  // Ordena diretamente
    }




    public double computeHeuristic(Ilayout goalLayout) {
        Platform current = (Platform) this;
        Platform goal = (Platform) goalLayout;

        double totalEstimatedCost = 0;

        if (cachedGoalPositions == null) {
            cachedGoalPositions = new HashMap<>();
            for (int i = 0; i < goal.stacks.size(); i++) {
                Stack<Container> goalStack = goal.stacks.get(i);
                for (int j = 0; j < goalStack.size(); j++) {
                    cachedGoalPositions.put(goalStack.get(j).getId(), new int[]{i, j});
                }
            }
        }

        for (int i = 0; i < current.stacks.size(); i++) {
            Stack<Container> currentStack = current.stacks.get(i);
            for (int j = 0; j < currentStack.size(); j++) {
                Container currentContainer = currentStack.get(j);
                int[] goalPosition = cachedGoalPositions.get(currentContainer.getId());

                // Ignora se já está na pilha e posição corretas
                if (goalPosition != null && !(j == 0 && goalPosition[1] == 0) && !(j > 0 && goalPosition[1] == j)) {
                    totalEstimatedCost += currentContainer.getCost();
                }
            }
        }

        return totalEstimatedCost;
    }











    /**
     * move o container de uma stack e coloca noutra stack completamente nova
     * @param fromStack representa a stack de onde será removido o container
     */
    private void moveToGround(int fromStack) {
        List<Stack<Container>> stacksLocal = getStacks();
        if (fromStack >= stacksLocal.size() || stacksLocal.get(fromStack).isEmpty()) return;

        Container container = stacksLocal.get(fromStack).pop();

        Stack<Container> newStack = new Stack<>();
        newStack.push(container);
        stacksLocal.add(newStack);

        this.setCost(container.getCost());
    }

    /**
     *
     * @param fromStack representa a stack de onde será removido o contêiner
     * @param toStack representa a stack para onde será enviado o contêiner
     */
    private void moveToStack(int fromStack, int toStack) {
        List<Stack<Container>> stacksLocal = getStacks();
        if (fromStack >= stacksLocal.size() || toStack >= stacksLocal.size() || stacksLocal.get(fromStack).isEmpty()) {
            return;
        }

        Container container = stacksLocal.get(fromStack).pop();
        stacksLocal.get(toStack).push(container);

        this.setCost(container.getCost());
    }
}

File: Inteligencia_Artificial_Programa_2C++/Platform.cpp
#ifndef PLATFORM_H
#define PLATFORM_H


#include <iostream>
#include <vector>
#include <stack>
#include <string>
#include <unordered_map>
#include <memory>
#include <algorithm> 
 // For std::sort
class Ilayout {
public:
    virtual double computeHeuristic(std::shared_ptr<Ilayout> goalLayout) = 0;
    virtual bool isGoal(std::shared_ptr<Ilayout> l) = 0;
    virtual std::vector<std::shared_ptr<Ilayout>> children() = 0;
    virtual double getK() const = 0;  // Movement cost
    virtual bool operator==(const Ilayout& other) const = 0;  // For equality check
    virtual std::size_t hashCode() const = 0;  // For hashing
    virtual ~Ilayout() {}
};

class Platform : public Ilayout {
private:
    std::vector<std::stack<Container>> stacks;  // Vector de pilhas de containers
    double cost;                                // Custo acumulado de movimentos
    std::unordered_map<char, std::pair<int, int>> cachedGoalPositions;  // Cache das posições objetivo

public:
    // Construtor padrão
    Platform() : cost(0) {}

    // Construtor com uma string de configuração
    Platform(const std::string& config, bool isInitialState) : cost(0) {
        parseInput(config, isInitialState);
    }

    // Construtor de cópia
    Platform(const Platform& other) : stacks(other.stacks), cost(other.cost), cachedGoalPositions(other.cachedGoalPositions) {}

    // Parse de uma configuração de entrada
    void parseInput(const std::string& config, bool isInitialState) {
        stacks.clear();  // Limpar as pilhas anteriores
        std::vector<std::string> stackConfigs;
        std::stringstream ss(config);
        std::string stackStr;

        // Dividir a string de configuração pelos espaços (representando as pilhas)
        while (ss >> stackStr) {
            std::stack<Container> stack;
            for (size_t i = 0; i < stackStr.length(); i++) {
                char containerId = stackStr[i];
                int moveCost = 0;

                // Verifica se há um número de custo de movimento ao lado do contêiner
                if (isInitialState && i + 1 < stackStr.length() && isdigit(stackStr[i + 1])) {
                    moveCost = stackStr[i + 1] - '0';  // Converte o dígito char para int
                    i++;  // Pula o próximo caractere (o número)
                }
                stack.push(Container(containerId, moveCost));
            }
            stacks.push_back(stack);
        }
    }

    // Getter para as pilhas
    const std::vector<std::stack<Container>>& getStacks() const {
        return stacks;
    }

    // Setter para o custo de movimento
    void setCost(double cost) {
        this->cost = cost;
    }

    // Função hash (necessária para armazenar em unordered_map)
    std::size_t hashCode() const override {
        std::size_t hash = 7;
        for (const auto& stack : stacks) {
            if (!stack.empty()) {
                hash = 31 * hash + std::hash<char>{}(stack.top().getId());
            }
        }
        return hash;
    }

    // Clone method
    std::shared_ptr<Ilayout> clone() const override {
        return std::make_shared<Platform>(*this);
    }

    // Verifica se esta configuração é igual a outra
    bool operator==(const Ilayout& other) const override {
        const Platform& otherPlatform = dynamic_cast<const Platform&>(other);
        return stacks == otherPlatform.stacks;
    }

    // Método de heurística para o A* (cálculo do custo estimado)
    double computeHeuristic(std::shared_ptr<Ilayout> goalLayout) override {
        double totalEstimatedCost = 0;
        Platform* goal = dynamic_cast<Platform*>(goalLayout.get());

        // Cache das posições objetivo
        if (cachedGoalPositions.empty()) {
            for (size_t i = 0; i < goal->stacks.size(); i++) {
                const std::stack<Container>& goalStack = goal->stacks[i];
                std::stack<Container> tempStack = goalStack;  // Cópia para iteração
                int level = 0;
                while (!tempStack.empty()) {
                    cachedGoalPositions[tempStack.top().getId()] = {i, level};
                    tempStack.pop();
                    level++;
                }
            }
        }

        // Cálculo do custo estimado
        for (size_t i = 0; i < stacks.size(); i++) {
            std::stack<Container> currentStack = stacks[i];
            int level = 0;
            while (!currentStack.empty()) {
                Container currentContainer = currentStack.top();
                currentStack.pop();
                auto goalPosition = cachedGoalPositions[currentContainer.getId()];
                if (goalPosition.first != i || goalPosition.second != level) {
                    totalEstimatedCost += currentContainer.getCost();
                }
                level++;
            }
        }

        return totalEstimatedCost;
    }

    // Verifica se esta configuração é o objetivo
    bool isGoal(std::shared_ptr<Ilayout> l) override {
        return *this == *l;
    }

    // Gera filhos (sucessores)
    std::vector<std::shared_ptr<Ilayout>> children() override {
        std::vector<std::shared_ptr<Ilayout>> result;

        // Para cada pilha, mover o topo para todas as outras pilhas
        for (size_t i = 0; i < stacks.size(); i++) {
            if (stacks[i].empty()) continue;  // Ignora pilhas vazias
            for (size_t j = 0; j < stacks.size(); j++) {
                if (i == j) continue;  // Não mover para a mesma pilha
                Platform child = *this;  // Clone da plataforma atual
                Container topContainer = child.stacks[i].top();
                child.stacks[i].pop();
                child.stacks[j].push(topContainer);
                child.setCost(child.getK() + topContainer.getCost());
                result.push_back(std::make_shared<Platform>(child));
            }
        }
        return result;
    }

    // Ordena as pilhas com base nos identificadores dos contêineres
    void getSortedStacks() {
        std::sort(stacks.begin(), stacks.end(), [](const std::stack<Container>& a, const std::stack<Container>& b) {
            if (a.empty() || b.empty()) return false;  // Ignora pilhas vazias
            return a.top().getId() < b.top().getId();  // Ordena pelo ID do topo
        });
    }

    // Move o contêiner de uma pilha para uma nova pilha (representando o chão)
    void moveToGround(int fromStack) {
        if (fromStack >= stacks.size() || stacks[fromStack].empty()) return;

        Container container = stacks[fromStack].top();
        stacks[fromStack].pop();

        std::stack<Container> newStack;
        newStack.push(container);
        stacks.push_back(newStack);  // Adiciona uma nova pilha (representando o chão)

        this->setCost(container.getCost());
    }

    // Move o contêiner de uma pilha para outra pilha
    void moveToStack(int fromStack, int toStack) {
        if (fromStack >= stacks.size() || toStack >= stacks.size() || stacks[fromStack].empty()) return;

        Container container = stacks[fromStack].top();
        stacks[fromStack].pop();
        stacks[toStack].push(container);

        this->setCost(container.getCost());
    }

    // Retorna o custo de movimento acumulado (g(n))
    double getK() const override {
        return cost;
    }

    // Representação textual da plataforma
    std::string toString() const {
        std::string result;
        for (const auto& stack : stacks) {
            result += "[";
            std::stack<Container> tempStack = stack;  // Cópia para iteração
            while (!tempStack.empty()) {
                result += tempStack.top().toString();
                tempStack.pop();
                if (!tempStack.empty()) result += ", ";
            }
            result += "]\n";
        }
        return result;
    }
};

#endif